### DISCUSSION TITLE
Additional documentation

### DISCUSSION THREAD

**Discussion by Lillecarl** [houmain/keymapper#88] [Q&A]
Hi!

Thanks for this great project! 😄 I'm working on [NixOS](https://github.com/Lillecarl/nixos/blob/master/modules/nixos/keymapper.nix) and [home-manager](https://github.com/Lillecarl/nixos/blob/master/modules/hm/keymapper.nix) [modules](https://nixos.wiki/wiki/NixOS_Modules) that I'll try to upstream.

Some things I noticed missing from the docs that might be good for first time users like myself of keyboard remapping solutions:

### Layout switching
For those unaware, you have to switch input layouts on the new device that keymapper creates rather than your good old tested at-translated-set-2-keyboard (or whatever).

### udev rules
Since keymapper can start whenever, restart whenever it doesn't have stable device names. This can be resolved with udev rules. Might be useful for applications like tp-auto-kbbl which reads inputs to toggle backlight on ThinkPads.
```
# Link keymapper to named devices
ACTION=="add", KERNEL=="event*", ATTRS{name}=="Keymapper", RUN+="/usr/bin/ln -sf /dev/input/%k /dev/input/keymapper_kb"
ACTION=="add", KERNEL=="mouse*", ATTRS{name}=="Keymapper", RUN+="/usr/bin/ln -sf /dev/input/%k /dev/input/keymapper_mouse"
ACTION=="remove", KERNEL=="event*", ATTRS{name}=="Keymapper", RUN+="/usr//bin/unlink /dev/input/keymapper_kb"
ACTION=="remove", KERNEL=="mouse*", ATTRS{name}=="Keymapper", RUN+="/usr/bin/unlink /dev/input/keymapper_mouse"
```
I guess keymapper could implement this linking within itself too if it'd want, but udev does the trick.

### Layers
There's not much regarding how one could implement layers. I'm using this configuration to achieve something resembling layers:
https://github.com/Lillecarl/nixos/blob/master/lillecarl/terminal/keymapper.nix#L7

That's all I've got, thanks for all your effort 😄

**Comment by houmain** [comment]
Hi Carl, 
thanks for your kind words and your work on simplifying the installation process on NixOS!
Good to know, how to get stable device names.
It is always interesting to see which configurations other people come up with. Maybe I should add a contribution notice or something to ask people to post their inventions to a discussion thread.

**Reply by Lillecarl** [reply]
That could indeed be useful, I had to mess about for a bit to figure out the device name rules 😄

**Reply by Zireael07** [reply]
The link to your config doesn't work @Lillecarl

**Reply by Lillecarl** [reply]
@Zireael07 https://github.com/Lillecarl/nixos/blob/a37fe1dc0b4ca737ae8817b86960e48f236d168e/modules/hm/keymapper.nix it's in the history.

I've come to the conclusion that I want to write my own remapping logic in Python so I'm not using keymapper anymore, but if you need a hand I'm glad to assist!

**Reply by ristomatti** [reply]
Just a side note, on line https://github.com/Lillecarl/nixos/blob/a37fe1dc0b4ca737ae8817b86960e48f236d168e/modules/hm/keymapper.nix#L39, `keymapper` is started with `keymapper -v -c FILE`, but in the repo's autostart script `extra/xdg/autostart/keymapper.desktop` it's started with `keymapper -u`, in to be able to modify the config on the fly. But I don't know anything about Nix, perhaps it would be against the convention...
